src/pyrokinetics/equilibrium/imas.py

```diff
@@ -0,0 +1,237 @@
+                psi_grid_new = np.concatenate((psi_grid[:lcfs_idx], [psi_lcfs_new]))
+                # Linearly interpolate each grid onto the new psi_grid
+                # Need psi to be increasing for np.interp
+                F = np.interp(psi_grid_new, psi_grid[::index], F[::index])
+                FF_prime = np.interp(psi_grid_new, psi_grid[::index], FF_prime[::index])
+                p = np.interp(psi_grid_new, psi_grid[::index], p[::index])
+                p_prime = np.interp(psi_grid_new, psi_grid[::index], p_prime[::index])
+                q = np.interp(psi_grid_new, psi_grid[::index], q[::index])
+                # Replace psi_grid and psi_lcfs with the new versions
+                psi_grid = psi_grid_new
+                psi_lcfs = psi_lcfs_new
+
+            # r_major, r_minor, and z_mid are not provided in the file. They must be
+            # determined by fitting contours to the psi_rz grid.
+            R_major = np.empty(len(psi_grid)) * units.meter
+            r_minor = np.empty(len(psi_grid)) * units.meter
+            Z_mid = np.empty(len(psi_grid)) * units.meter
+            R_major[0] = R_axis
+            r_minor[0] = 0.0 * units.meter
+            Z_mid[0] = Z_axis
+
+            Rbdry = data["time_slice[]&boundary&outline&r"][time_index, :]
+            Zbdry = data["time_slice[]&boundary&outline&z"][time_index, :]
+
+            R_major[-1] = 0.5 * (max(Rbdry) + min(Rbdry)) * units.meter
+            r_minor[-1] = 0.5 * (max(Rbdry) - min(Rbdry)) * units.meter
+            Z_mid[-1] = 0.5 * (max(Zbdry) + min(Zbdry)) * units.meter
+            for idx, psi in enumerate(psi_grid[1:-1], start=1):
+                Rc, Zc = _flux_surface_contour(R, Z, psi_RZ, R_axis, Z_axis, psi)
+                R_min, R_max = min(Rc), max(Rc)
+                Z_min, Z_max = min(Zc), max(Zc)
+                R_major[idx] = 0.5 * (R_max + R_min)
+                r_minor[idx] = 0.5 * (R_max - R_min)
+                Z_mid[idx] = 0.5 * (Z_max + Z_min)
+
+            a_minor = r_minor[-1]
+
+        # Create and return Equilibrium
+        return Equilibrium(
+            R=R,
+            Z=Z,
+            psi_RZ=psi_RZ,
+            psi=psi_grid,
+            F=F,
+            FF_prime=FF_prime,
+            p=p,
+            p_prime=p_prime,
+            q=q,
+            R_major=R_major,
+            r_minor=r_minor,
+            Z_mid=Z_mid,
+            psi_lcfs=psi_lcfs,
+            a_minor=a_minor,
+            B_0=B_0,
+            I_p=I_p,
+            clockwise_phi=clockwise_phi,
+            cocos=cocos,
+            eq_type="IMAS",
+        )
+
+    def verify_file_type(self, filename: PathLike) -> None:
+        """Quickly verify that we're looking at a IMAS file without processing"""
+        # Try opening the IMAS file using freeqdsk.geqdsk
+        filename = Path(filename)
+        if not filename.is_file():
+            raise FileNotFoundError(filename)
+        try:
+            raw_data = h5py.File(filename, "r")
```

Comment:
I'm a little worried this technically leaves the file open. Maybe this instead?

```suggestion
            with h5py.File(filename, "r") as f:
                keys = f.keys()
```

Reply:
Yes I think you're right